meetings/working-groups/params-improvements/PI-2022-11-03.md

```diff
@@ -0,0 +1,126 @@
+
+## Span allocation
+The compiler _may allocate a span on the stack_ when:
+- the span is implicitly allocated,
+- the span length is known at compile time, and
+- the compiler can determine _through escape analysis_ that no references to the span escape the containing method.
+
+For arguments to a `params ReadOnlySpan<T>` parameter (which is implicitly `scoped`), the conditions above are satisfied.
+
+We have a request with the runtime team to support _fixed size arrays_ of managed types, at least for fields.
+If we have _fixed size array_ fields, we can define `struct` types with inline arrays and use locals for stack allocated arrays.
+
+For example, consider a `FixedSizeArray2` type defined below which includes an inline 2 element array:
+```csharp
+struct FixedSizeArray2<T>
+{
+    public T[2] Array; // pseudo-code for inline array
+}
+```
+
+With that type, a call to `WriteLine("{0}, {1}", x, y)` could be emitted as:
+```csharp
+var _tmp1 = new FixedSizeArray2<object>();
+_tmp1.Array[0] = x;
+_tmp1.Array[1] = y;
+var _tmp2 = new ReadOnlySpan<object>(_tmp.Array);
+
+// WriteLine(string format, params ReadOnlySpan<object> args);
+WriteLine("{0}, {1}", _tmp2);
+```
+
+Ideally the BCL will provide types such as `FixedSizeArray1`, `FixedSizeArray2`, etc. for a limited number of array lengths.
+And if the compilation requires spans for other argument lengths, the compiler would generate and emit the additional types.
+
+The compiler will implicitly allocate `params` argument spans on the stack regardless of the number of arguments.
+To avoid implicit stack allocation at a call site, the calling code should allocate the span explicitly.
+
+_What is the fallback for other target frameworks: heap allocation or compiler-generated types with contiguous fields: `struct FixedSizeArrayN { T T0; ... T TN; }`?_
+
+## Span reuse
+The compiler _may reuse a span_ when:
+- the span is implicitly allocated,
+- the span length is known at compile time,
+- the compiler can determine _through escape analysis_ that no references to the span escape the containing method, and
+- the compiler can determine there are _no aliases_ to the span in user code.
+
+For arguments to a `params ReadOnlySpan<T>` parameter (which is implicitly `scoped`), the conditions above are satisfied.
+
+The compiler will reuse a single span across all calls to `params ReadOnlySpan<T>` methods where the span element types are considered identical by the runtime.
+
+The length of the reused span will be the length of the longest `params` argument list from all uses.
+(The actual span argument passed to a `params` method will be a slice of that reused span, with the expected length for the particular call.)
+
+Reuse may be across distinct call sites _or_ repeated calls from the same call site.
+Reuse is per thread of execution and within the same method only.
+No reuse across lambda expressions, local functions, and the containing methods.
```

Comment:
the local function case seems unfortunate, esp if we can see that it is never captured as a delegate.  could we distinguish those cases and allow reuse across local function for when that doesn't happen?

Reply:
The cost of these allocations (assuming they end up on the stack) is likely to be exceedingly low. Possibly just the cost of zero'ing the memory region. Of course this needs measurement to be absolutely confident, but it shouldn't be all that expensive.

Reply:
It sounds like this would just be a stack space optimization. I think the compiler should be permitted to perform such an optimization, but that we should punt out actually doing it until somebody needs it. I could imagine it getting difficult to do the optimization correctly in reentrant scenarios, i.e. where local functions have `params ROS` parameters and call into themselves or each other.

Reply:
i agree with @RikkiGibson .  I think it's too strong to say "no reuse". I think it would be better to say that it is permitted.